File: 965.单值二叉树.go
/*
 * @lc app=leetcode.cn id=965 lang=golang
 *
 * [965] 单值二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// 执行用时：0 ms, 在所有 Go 提交中击败了100.00%的用户
// 内存消耗：2.1 MB, 在所有 Go 提交中击败了45.83%的用户
func isUnivalTree(root *TreeNode) bool {
    preNum := root.Val  // 记录根节点的值，后续不相同直接返回
    var dfs func(node *TreeNode) bool
    dfs = func(node *TreeNode) bool {
        if node == nil {    // 正常结束
            return false
        } else if node.Val != preNum {  // 不相等返回
            return true
        }
        // 判定左右子节点
        return dfs(node.Left) || dfs(node.Right)
    }
    // 根节点的左右子节点判定开始即可
    if dfs(root.Left) || dfs(root.Right) {  // 其中有一个不相同
        return false
    }
    // 满足单值二叉树定义
    return true
}
// @lc code=end

File: 965.单值二叉树.py
#
# @lc app=leetcode.cn id=965 lang=python3
#
# [965] 单值二叉树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isUnivalTree(self, root: TreeNode) -> bool:
        pre_num = root.val

        def dfs(node: TreeNode) -> bool:
            if not node:    # 空则正常
                return False
            if node.val != pre_num:     # 不相等直接退出
                return True
            # 目前为止满足单值二叉树定义
            return dfs(node.left) or dfs(node.right)
        if dfs(root.left) or dfs(root.right):
            # 不满足条件
            return False
        return True
# @lc code=end

